File: solutions/java/0396.java
class Solution {
  public int maxRotateFunction(int[] A) {
    int sum = Arrays.stream(A).sum();
    int f = 0;

    for (int i = 0; i < A.length; ++i)
      f += i * A[i];

    int ans = f;

    for (int i = A.length - 1; i >= 0; --i) {
      f += sum - A.length * A[i];
      ans = Math.max(ans, f);
    }

    return ans;
  }
}

File: solutions/cpp/0396.cpp
class Solution {
 public:
  int maxRotateFunction(vector<int>& A) {
    const int sum = accumulate(begin(A), end(A), (long)0);
    int f = 0;

    for (int i = 0; i < A.size(); ++i)
      f += i * A[i];  // calculate F(0) first

    int ans = f;

    for (int i = A.size() - 1; i > 0; --i) {
      f += sum - A.size() * A[i];
      ans = max(ans, f);
    }

    return ans;
  }
};